src/software/unix_full_system_main.cpp

```diff
@@ -14,12 +14,32 @@
 #include "software/constants.h"
 #include "software/estop/arduino_util.h"
 #include "software/logger/logger.h"
+#include "software/logger/proto_logger.h"
 #include "software/multithreading/observer_subject_adapter.hpp"
 #include "software/networking/udp/threaded_proto_udp_listener.hpp"
 #include "software/networking/unix/threaded_proto_unix_listener.hpp"
 #include "software/sensor_fusion/threaded_sensor_fusion.h"
 #include "software/util/generic_factory/generic_factory.h"
 
+// ProtoLogger has to be defined as a global variable so that it can be accessed by the
+// cleanup callback.
+std::shared_ptr<ProtoLogger> proto_logger;
+
+/**
+ * Signal handler to cleanly shutdown the program
+ * @param signal_num The signal number that was caught
+ */
+void cleanup(int signal_num)
+{
+    if (proto_logger)
+    {
+        proto_logger->flushAndStopLogging();
+    }
+
+    // Program has cleaned up core resources, so we can safely exit
+    exit(0);
+}
```

Comment:
This signal handler calls some functions that are not signal-safe (pretty much all of the standard library is unsafe including `std::cout`, `exit`, smart pointers). We can't guarantee that everything will work (except reentrant functions) bc technically we don't know what the state of our program is. I don't know if this will actually cause any problems in reality, but maybe we should note this in a comment.

Reply:
This is interesting. I wasn't familiar with signal-safe functions. I've added a comment from the little bit I understand, but feel free to let me know if I can improve it.